pkg/block/indexheader/header_test.go

```diff
@@ -20,53 +31,362 @@ func TestReaders(t *testing.T) {
+	   }
+	   app.Commit()
+	   db.compact()
+	   db.Close()
+	*/
+
+	m, err := metadata.Read("./testdata/index_format_v1")
+	testutil.Ok(t, err)
+	testutil.Copy(t, "./testdata/index_format_v1", filepath.Join(tmpDir, m.ULID.String()))
+
+	_, err = metadata.InjectThanos(log.NewNopLogger(), filepath.Join(tmpDir, m.ULID.String()), metadata.Thanos{
+		Labels:     labels.Labels{{Name: "ext1", Value: "1"}}.Map(),
+		Downsample: metadata.ThanosDownsample{Resolution: 0},
+		Source:     metadata.TestSource,
+	}, &m.BlockMeta)
+	testutil.Ok(t, err)
+	testutil.Ok(t, block.Upload(ctx, log.NewNopLogger(), bkt, filepath.Join(tmpDir, m.ULID.String())))
+
+	for _, id := range []ulid.ULID{id1, m.ULID} {
+		t.Run(id.String(), func(t *testing.T) {
+			indexFile, err := fileutil.OpenMmapFile(filepath.Join(tmpDir, id.String(), block.IndexFilename))
+			testutil.Ok(t, err)
+			defer func() { _ = indexFile.Close() }()
+
+			b := realByteSlice(indexFile.Bytes())
+
+			t.Run("binary", func(t *testing.T) {
+				fn := filepath.Join(tmpDir, id.String(), block.IndexHeaderFilename)
+				testutil.Ok(t, WriteBinary(ctx, bkt, id, fn))
+
+				br, err := NewBinaryReader(ctx, log.NewNopLogger(), nil, tmpDir, id)
+				testutil.Ok(t, err)
+
+				defer func() { testutil.Ok(t, br.Close()) }()
+
+				if id == id1 {
+					testutil.Equals(t, 1, br.version)
+					testutil.Equals(t, 2, br.indexVersion)
+					testutil.Equals(t, &BinaryTOC{Symbols: headerLen, PostingsOffsetTable: 50}, br.toc)
+					testutil.Equals(t, int64(330), br.indexLastPostingEnd)
+					testutil.Equals(t, 8, br.symbols.Size())
+					testutil.Equals(t, 3, len(br.postings))
+					testutil.Equals(t, 0, len(br.postingsV1))
+					testutil.Equals(t, 2, len(br.nameSymbols))
+				}
 
-		testutil.Equals(t, 6, len(jr.symbols))
-		testutil.Equals(t, 2, len(jr.lvals))
-		testutil.Equals(t, 6, len(jr.postings))
+				compareIndexToHeader(t, b, br)
+			})
+
+			t.Run("json", func(t *testing.T) {
+				fn := filepath.Join(tmpDir, id.String(), block.IndexCacheFilename)
+				testutil.Ok(t, WriteJSON(log.NewNopLogger(), filepath.Join(tmpDir, id.String(), "index"), fn))
+
+				jr, err := NewJSONReader(ctx, log.NewNopLogger(), nil, tmpDir, id)
+				testutil.Ok(t, err)
+
+				defer func() { testutil.Ok(t, jr.Close()) }()
+
+				if id == id1 {
+					testutil.Equals(t, 6, len(jr.symbols))
+					testutil.Equals(t, 2, len(jr.lvals))
+					testutil.Equals(t, 6, len(jr.postings))
+				}
+
+				compareIndexToHeader(t, b, jr)
+			})
+		})
+	}
 
-		testReader(t, jr)
-	})
 }
 
-func testReader(t *testing.T, r Reader) {
-	testutil.Equals(t, 2, r.IndexVersion())
-	exp := []string{"1", "2", "3", "4", "a", "b"}
-	for i := range exp {
-		r, err := r.LookupSymbol(uint32(i))
+func compareIndexToHeader(t *testing.T, indexByteSlice index.ByteSlice, headerReader Reader) {
+	indexReader, err := index.NewReader(indexByteSlice)
+	testutil.Ok(t, err)
+	defer func() { _ = indexReader.Close() }()
```

Comment:
Shouldn't we check that this succeeds?

Reply:
no, because it's a Prometheus indexReader and we are not testing that one.